src/Features/LanguageServer/Protocol/Handler/RequestContext.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    /// <summary>
+    /// Context for requests handled by <see cref="IRequestHandler"/>
+    /// </summary>
+    internal struct RequestContext
+    {
+        /// <summary>
+        /// The client capabilities for the request.
+        /// </summary>
+        public ClientCapabilities ClientCapabilities { get; }
+
+        /// <summary>
+        /// The LSP client making the request
+        /// </summary>
+        public string? ClientName { get; }
+
+        /// <summary>
+        /// A cancellation token
+        /// </summary>
+        public CancellationToken CancellationToken { get; }
```

Comment:
nit - I think I would prefer for this to be left as an explicit parameter to the request, that way it's very easy to see the cancellation chain from the method calls.

not sure what best practice is though for this, so happy to be proved wrong!

Reply:
I was in two minds about this too, but ended up coping CodeFixContext. Will see if I can work out which way is idiomatic.

Reply:
we have precedence for both.  However, i lean toward B's view on this.  Specifically because we have had bugs where someone does thsi:

```
void FOo(SomethingContext context, CancellationToken cancellationToken)
{
     // which token do i use?
}
```

So my personal pref is to not combine them.  But i do understand the ship has sailed with some very public APIs.

Reply:
(It's a reason i also don't like storing teh cancellation token as a field like `_cancellationToken`.  Def had had circumstances where the wrong one was used, or there was ambiguity/confusion about what to pass along.

Reply:
Given that I fell into the trap of not using a cancellation token, because it was hiding on the CodeFixContext, I'm glad to know thats not the norm. I've promoted it back out to a parameter.